nvflare/app_common/utils/fl_model_utils.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvflare.apis.dxo import DXO, DataKind, from_shareable
+from nvflare.apis.shareable import Shareable
+from nvflare.app_common.abstract.fl_model import FLModel, FLModelConst, TransferType
+from nvflare.app_common.app_constant import AppConstants
+
+MODEL_ATTRS = [
+    FLModelConst.MODEL,
+    FLModelConst.METRICS,
+    FLModelConst.TRANSFER_TYPE,
+    FLModelConst.OPTIMIZER,
+    FLModelConst.METRICS,
+    FLModelConst.CONFIGS,
+    FLModelConst.CLIENT_WEIGHTS,
+    FLModelConst.ROUND,
+    FLModelConst.TOTAL_ROUNDS,
+    FLModelConst.META,
+]
+
+
+transfer_type_to_data_kind = {
+    TransferType.MODEL.value: DataKind.WEIGHTS,
+    TransferType.MODEL_DIFF.value: DataKind.WEIGHT_DIFF,
+    TransferType.METRICS_ONLY.value: DataKind.METRICS,
+}
```

Comment:
I thought we were going to make TT and DK 1:1 for:
TT.MODEL <=> DK.MODEL
TT.MODEL_DIFF <=> DK.MODEL_DIFF
TT.WEIGHTS <=> DK.WEIGHTS
TT.WEIGHT_DIFF <=> DK.WEIGHT_DIFF

I'm not sure about TT.METRICS_ONLY. Why do we need this? Everything in FLModel is optional anyway.

Reply:
Instead of "Metrics_only", what about "non_model" to cover all other stuff including metric / config / just-some-random-meta?  @chesterxgchen for https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_common/abstract/fl_model.py

Reply:
I feel the same way. Or just make transfer_type default to None. It only specifies the nature of the "model".

Reply:
I was making it 1-to-1, say TT.MODEL -> DK.MODEL.

But then I found that on server side we have components that check for the DataKind needs to be Weights or Weights_diff.

That's why I went back to TT.MODEL -> DK.WEIGHTS

But I guess we could just make it 1-to-1 and if any components on server side not supporting these new data kind then we will need to update them later.

Reply:
For example, we will need to update the following places afterwards:

./nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py
./nvflare/app_common/aggregators/dxo_aggregator.py

If want to support HE:

./nvflare/app_opt/he/model_encryptor.py
./nvflare/app_opt/he/model_shareable_generator.py
./nvflare/app_opt/he/model_decryptor.py

I think that's good as MODEL, MODEL_DIFF is more general than WEIGHT, WEIGHT_DIFF.
We should make those components more general.

Reply:
If "non_model to cover all other stuff including metric / config / just-some-random-meta", should they be using DXO?

I don't know how general we want this FLModel to be.

Reply:
We could call it `FLData` to cover both models and metrics.

Reply:
Might make sense to separate these concepts into `FLModel(FLData)` & `FLMetrics(FLData)`

Reply:
If we have

TT.MODEL
TT.MODEL_DIFF
TT.WEIGHTS 
TT.WEIGHT_DIFF 
TT.METRICS

Then maybe this name "transfer_type" is not suitable anymore?
It should be "model_type"?

Reply:
Not really. FLData would have all the args we have now. Just that FLMetrics wouldn't have the model related fields like weights and optimizer. Maybe `FLModel(FLMetrics)` is the better way to go.

Reply:
thanks for the thoughts, let's continue that discussion in a separate PR to enhance/enlarge this scope.